picture-stealer/src/main/java/com/bobocode/picturestealer/service/PhotoService.java

```diff
@@ -0,0 +1,46 @@
+package com.bobocode.picturestealer.service;
+
+import com.bobocode.picturestealer.dto.Photo;
+import com.bobocode.picturestealer.entity.CameraEntity;
+import com.bobocode.picturestealer.entity.PhotoEntity;
+import com.bobocode.picturestealer.manager.PhotoRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author "Maksym Oliinyk"
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class PhotoService {
+
+    private final CameraService cameraService;
+    private final PhotoRepository photoRepository;
+
+    //todo: add retry mechanism for resolve OptimisticLockException in case of concurrent requests
+    @Transactional(propagation = Propagation.NOT_SUPPORTED)
```

Comment:
@WHALE88 please consider the following code example, as an alternative way of writing the same code:

```
    public PhotoEntity createIfNotExists(Photo photo) {
        log.debug("Retrieve PhotoEntity by nasa identifier {}", photo.id());
        return photoRepository.findByNasaId(photo.id()))
                              .orElseGet(() -> createNewPhotoEntity(photo));
    }

    public PhotoEntity createNewPhotoEntity(Photo photo) {
        log.info("No PhotoEntity found by nasa identifier {}, so creating a new entity", photo.id());
        final CameraEntity cameraEntity = cameraService.findOrCreateCamera(photo.camera());
        final PhotoEntity photoEntity = new PhotoEntity(photo.id(), cameraEntity, photo.imgSrc())
        return photoRepository.save(photoEntity);
    }
```

Reply:
@tboychuk Absolutely agree with you